pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go

```diff
@@ -0,0 +1,185 @@
+	// return true if the created namespace is selected for discovery
+	NamespaceCreated(ns metav1.ObjectMeta) (membershipChanged bool)
+	// membershipChanged will be true if the updated namespace is newly selected or deselected for discovery
+	NamespaceUpdated(oldNs, newNs metav1.ObjectMeta) (membershipChanged bool, namespaceAdded bool)
+	// return true if the deleted namespace was selected for discovery
+	NamespaceDeleted(ns metav1.ObjectMeta) (membershipChanged bool)
+}
+
+type discoveryNamespacesFilter struct {
+	lock                sync.RWMutex
+	nsLister            listerv1.NamespaceLister
+	discoveryNamespaces sets.String
+	discoverySelectors  []labels.Selector // nil if discovery selectors are not specified, permits all namespaces for discovery
+}
+
+func NewDiscoveryNamespacesFilter(
+	nsLister listerv1.NamespaceLister,
+	discoverySelectors []*metav1.LabelSelector,
+) DiscoveryNamespacesFilter {
+	discoveryNamespacesFilter := &discoveryNamespacesFilter{
+		nsLister: nsLister,
+	}
+
+	// initialize discovery namespaces filter
+	discoveryNamespacesFilter.SelectorsChanged(discoverySelectors)
+
+	return discoveryNamespacesFilter
+}
+
+func (d *discoveryNamespacesFilter) Filter(obj interface{}) bool {
+	// permit all objects if discovery selectors are not specified
+	if len(d.discoverySelectors) == 0 {
+		return true
+	}
+
+	// permit if object resides in a namespace labeled for discovery
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	return d.discoveryNamespaces.Has(obj.(metav1.Object).GetNamespace())
+}
+
+// initialize the discovery filter state with the discovery selectors and selected namespaces
+func (d *discoveryNamespacesFilter) SelectorsChanged(
+	discoverySelectors []*metav1.LabelSelector,
+) (selectedNamespaces []string, deselectedNamespaces []string) {
+	oldDiscoveryNamespaces := d.discoveryNamespaces
+
+	var selectors []labels.Selector
+	newDiscoveryNamespaces := sets.NewString()
+
+	namespaceList, err := d.nsLister.List(labels.Everything())
+	if err != nil {
+		log.Errorf("error initializing discovery namespaces filter, failed to list namespaces: %v", err)
+		return
+	}
+
+	// convert LabelSelectors to Selectors
+	for _, selector := range discoverySelectors {
+		ls, err := metav1.LabelSelectorAsSelector(selector)
+		if err != nil {
+			log.Errorf("error initializing discovery namespaces filter, invalid discovery selector: %v", err)
+			return
+		}
+		selectors = append(selectors, ls)
+	}
+
+	if len(selectors) != 0 {
+		// range over all namespaces to get discovery namespaces
+		for _, ns := range namespaceList {
+			for _, selector := range selectors {
+				// omitting discoverySelectors indicates discovering all namespaces
+				if selector.Matches(labels.Set(ns.Labels)) {
+					newDiscoveryNamespaces.Insert(ns.Name)
+				}
+			}
+		}
+	}
+
+	selectedNamespaces = newDiscoveryNamespaces.Difference(oldDiscoveryNamespaces).List()
```

Comment:
this also needs a lock. probably the same one as below, although this could technically be a rlock

Reply:
i'll use the same `Lock` for this and below